compiler/rustc_mir_build/src/thir/pattern/mod.rs

```diff
@@ -755,11 +757,21 @@ pub(crate) fn compare_const_vals<'tcx>(
 ) -> Option<Ordering> {
     assert_eq!(a.ty(), b.ty());
 
-    if a == b {
-        return Some(Ordering::Equal);
+    let ty = a.ty();
+
+    // This code is hot when compiling `unicode-normalization` because it has a
+    // number of matches with many ranges such as '\u{037A}'..='\u{037F}'. So
+    // we special-case comparisons of chars in the relevant form for speed.
```

Comment:
I dislike special casing chars just for one benchmark. I think we should generally improve the logic inside the `eval_bits` and related functions to have a fast path for already evaluated constants.

For something that is actionable in this PR, please just replace this specific type match with one that performs this fast path for everything but floats and signed ints (or possibly pull the large type match below into a separate inline-always function and call it whenever both constants are already evaluated, irrespective of the type.

Reply:
Creating a fast path like that is tricky, because there are so many layers to these types, and because of the size checking and layout computations that takes place.

The fast path I added to this function hard to integrate with the rest of the function, because the fast path comparison is on `ScalarInt` values, while the general case comparison is on `u128` values. Extracting those `u128` values in the fast case would require extra size checks and layout computations, eating into the speed wins. But I have generalized it to work with any scalar type other than `Float` or `Int`.